File: android/app/src/main/java/de/kjell/zencompanion/sync/HawkAuth.kt
package de.kjell.zencompanion.sync

import java.net.URL
import java.security.SecureRandom
import java.util.Base64

/**
 * Port of `Shared/HawkAuth.swift`. Signs the verbatim request-line resource
 * (percent-encoding preserved), exactly like the Swift implementation.
 */
object HawkAuth {
    private val random = SecureRandom()

    fun authorization(
        method: String,
        url: URL,
        id: String,
        key: ByteArray,
        payloadHash: String? = null,
        resource: String? = null,
        fixedTimestamp: Long? = null,
        fixedNonce: String? = null,
    ): String {
        val ts = fixedTimestamp ?: (System.currentTimeMillis() / 1000)
        val nonce = fixedNonce ?: randomNonce()
        val host = url.host
        val port = if (url.port != -1) url.port else if (url.protocol == "https") 443 else 80
        // Swift's `URL.path` decodes percent-escapes, but Hawk must sign the
        // exact encoded resource the request line carries. Callers that build
        // paths with percent-encoded segments pass `resource` explicitly.
        //
        // On Android, `URL.path` keeps escapes as written; to stay behaviorally
        // identical with the Swift call sites we still prefer an explicit
        // `resource`, and otherwise use the path + query verbatim.
        val path: String = if (resource != null) {
            if (resource.startsWith("/")) resource else "/$resource"
        } else {
            val p = url.path.ifEmpty { "/" }
            val q = url.query
            if (!q.isNullOrEmpty()) "$p?$q" else p
        }
        val hash = payloadHash ?: ""
        val ext = ""
        val normalized = listOf(
            "hawk.1.header",
            ts.toString(),
            nonce,
            method.uppercase(),
            path,
            host.lowercase(),
            port.toString(),
            hash,
            ext,
            "",
        ).joinToString("\n")
        val macB64 = macFor(normalized, key)
        val parts = mutableListOf(
            "id=\"$id\"",
            "ts=\"$ts\"",
            "nonce=\"$nonce\"",
            "mac=\"$macB64\"",
        )
        if (hash.isNotEmpty()) parts.add("hash=\"$hash\"")
        return "Hawk " + parts.joinToString(", ")
    }

    fun macFor(normalized: String, key: ByteArray): String =
        Base64.getEncoder().encodeToString(FxACrypto.hmacSHA256(key, normalized.toByteArray(Charsets.UTF_8)))

    fun payloadHash(body: ByteArray, contentType: String = "application/json"): String {
        val normalizedType = contentType.split(";").firstOrNull()?.trim() ?: contentType
        val data = "hawk.1.payload\n$normalizedType\n".toByteArray(Charsets.UTF_8) +
            body + byteArrayOf(0x0A)
        return Base64.getEncoder().encodeToString(FxACrypto.sha256(data))
    }

    private fun randomNonce(): String {
        val bytes = ByteArray(8)
        random.nextBytes(bytes)
        return Base64.getEncoder().encodeToString(bytes)
            .replace("+", "")
            .replace("/", "")
            .replace("=", "")
    }
}

File: ios/Shared/Sync/HawkAuth.swift
import CryptoKit
import Foundation

enum HawkAuth {
    static func authorization(
        method: String,
        url: URL,
        id: String,
        key: Data,
        extraHeaders: [String: String] = [:],
        payloadHash: String? = nil,
        resource: String? = nil,
        fixedTimestamp: Int? = nil,
        fixedNonce: String? = nil
    ) -> String {
        let ts = fixedTimestamp ?? Int(Date().timeIntervalSince1970)
        let nonce = fixedNonce ?? randomNonce()
        let host = url.host ?? ""
        let port = url.port ?? (url.scheme == "https" ? 443 : 80)
        // `URL.path` decodes percent-escapes, but Hawk must sign the exact
        // encoded resource the request line carries. Callers that build paths
        // with percent-encoded segments pass `resource` explicitly.
        var path: String
        if let resource {
            path = resource.hasPrefix("/") ? resource : "/" + resource
        } else {
            path = url.path.isEmpty ? "/" : url.path
            if let query = url.query, !query.isEmpty {
                path += "?" + query
            }
        }
        let hash = payloadHash ?? ""
        let ext = ""
        let normalized = [
            "hawk.1.header",
            "\(ts)",
            nonce,
            method.uppercased(),
            path,
            host.lowercased(),
            "\(port)",
            hash,
            ext,
            ""
        ].joined(separator: "\n")
        let macB64 = macFor(normalized: normalized, key: key)
        var parts = [
            "id=\"\(id)\"",
            "ts=\"\(ts)\"",
            "nonce=\"\(nonce)\"",
            "mac=\"\(macB64)\""
        ]
        if !hash.isEmpty {
            parts.append("hash=\"\(hash)\"")
        }
        _ = extraHeaders
        return "Hawk " + parts.joined(separator: ", ")
    }

    static func macFor(normalized: String, key: Data) -> String {
        FxACrypto.hmacSHA256(key: key, data: Data(normalized.utf8)).base64EncodedString()
    }

    static func payloadHash(_ body: Data, contentType: String = "application/json") -> String {
        let normalizedType = contentType.split(separator: ";").first.map(String.init) ?? contentType
        var data = Data("hawk.1.payload\n\(normalizedType)\n".utf8)
        data.append(body)
        data.append(0x0A)
        return Data(SHA256.hash(data: data)).base64EncodedString()
    }

    private static func randomNonce() -> String {
        var bytes = [UInt8](repeating: 0, count: 8)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64EncodedString()
            .replacingOccurrences(of: "+", with: "")
            .replacingOccurrences(of: "/", with: "")
            .replacingOccurrences(of: "=", with: "")
    }
}
